http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -0,0 +1,441 @@
+            crtHeaderList.add(new HttpHeader(CONNECTION, KEEP_ALIVE));
+        }
+
+        // Set Content-Length if needed
+        Optional<Long> contentLength = asyncRequest.requestContentPublisher().contentLength();
+        if (isNullOrEmpty(sdkRequest.headers().get(CONTENT_LENGTH)) && contentLength.isPresent()) {
+            crtHeaderList.add(new HttpHeader(CONTENT_LENGTH, Long.toString(contentLength.get())));
+        }
+
+        // Add the rest of the Headers
+        for (Map.Entry<String, List<String>> headerList: sdkRequest.headers().entrySet()) {
+            for (String val: headerList.getValue()) {
+                HttpHeader h = new HttpHeader(headerList.getKey(), val);
+                crtHeaderList.add(h);
+            }
+        }
+
+        return crtHeaderList;
+    }
+
+    private HttpHeader[] asArray(List<HttpHeader> crtHeaderList) {
+        return crtHeaderList.toArray(new HttpHeader[crtHeaderList.size()]);
+    }
+
+    private HttpRequest toCrtRequest(URI uri, AsyncExecuteRequest asyncRequest, AwsCrtAsyncHttpStreamAdapter crtToSdkAdapter) {
+        SdkHttpRequest sdkRequest = asyncRequest.request();
+        Validate.notNull(uri, NULL_URI_ERROR_MESSAGE);
+        Validate.notNull(sdkRequest, NULL_REQUEST_ERROR_MESSAGE);
+
+        String method = sdkRequest.method().name();
+        String encodedPath = sdkRequest.encodedPath();
+        if (encodedPath == null || encodedPath.length() == 0) {
+            encodedPath = "/";
+        }
+
+        String encodedQueryString = SdkHttpUtils.encodeAndFlattenQueryParameters(sdkRequest.rawQueryParameters())
+                .map(value -> "?" + value)
+                .orElse("");
+
+        HttpHeader[] crtHeaderArray = asArray(createHttpHeaderList(uri, asyncRequest));
+
+        return new HttpRequest(method, encodedPath + encodedQueryString, crtHeaderArray, crtToSdkAdapter);
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(AsyncExecuteRequest asyncRequest) {
+
+        Validate.notNull(asyncRequest, "AsyncExecuteRequest must not be null");
+        Validate.notNull(asyncRequest.request(), NULL_REQUEST_ERROR_MESSAGE);
+        Validate.notNull(asyncRequest.requestContentPublisher(), "RequestContentPublisher must not be null");
+        Validate.notNull(asyncRequest.responseHandler(), "ResponseHandler must not be null");
+
+        URI uri = toUri(asyncRequest.request());
+
+        try (HttpClientConnectionManager crtConnPool = getOrCreateConnectionPool(uri)) {
+            CompletableFuture<Void> requestFuture = new CompletableFuture<>();
+
+            // When a Connection is ready from the Connection Pool, schedule the Request on the connection
+            crtConnPool.acquireConnection()
+                    .whenComplete((crtConn, throwable) -> {
+                        // If we didn't get a connection for some reason, fail the request
+                        if (throwable != null) {
```

Comment:
Looking at the subclasses of IoException and the ways this could fail I think wrapping this in an IoException is reasonable.  The most common problems would be socket in nature.